crates/llama-cpp-bindings/src/engine.cc

```diff
@@ -77,6 +84,39 @@ std::string string_format(const std::string& format, Args ... args)
 	return std::string(buf.get(), buf.get() + size - 1); // We don't want the '\0' inside
 }
 
+float compute_softmax_inplace(float* nums, size_t len, float temperature) {
+  float sum = 0;
+  float max = *std::max_element(nums, nums + len);
+  for (size_t i = 0; i < len; i++) {
+    nums[i] -= max;
+    nums[i] = std::exp(nums[i] / temperature);
+    sum += nums[i];
+  }
+  for (size_t i = 0; i < len; i++) {
+      nums[i] /= sum;
+  }
+  return sum;
+}
+
+size_t weighted_random(const float* nums, size_t len, uint64_t seed) {
+  std::mt19937 rng(seed);
+  float sum = 0;
```

Comment:
doesn't softmax above guranteed the `sum` being computed here ends up = `1.0`?

Reply:
make sense, while since we compute the sum in `compute_softmax_in_place`, let's just pass it here to avoid recomputing?